Userland/Libraries/LibWeb/HTML/Navigable.cpp

```diff
@@ -1152,57 +1140,48 @@ WebIDL::ExceptionOr<void> Navigable::populate_session_history_entry_document(
+        // FIXME: 5. Otherwise, if navigationParams's response has a `Content-Disposition`
         //            header specifying the attachment disposition type, then:
-        // 11. Otherwise:
-        else {
-            // 1. Let document be the result of loading a document given navigationParams, sourceSnapshotParams,
-            //    and entry's document state's initiator origin.
+        // 6. Otherwise, if navigationParams's response's status is not 204 and is not 205, then set entry's document state's document to the result of
+        //    loading a document given navigationParams, sourceSnapshotParams, and entry's document state's initiator origin.
+        else if (navigation_params.get<JS::NonnullGCPtr<NavigationParams>>()->response->status() != 204 && navigation_params.get<JS::NonnullGCPtr<NavigationParams>>()->response->status() != 205) {
             auto document = load_document(navigation_params.get<JS::NonnullGCPtr<NavigationParams>>());
-
-            // 2. If document is null, then run completionSteps and return.
-            if (!document) {
-                if (completion_steps)
-                    completion_steps->function()();
-                return;
-            }
-
-            // 3. Set entry's document state's document to document.
-            entry->document_state()->set_document(document.ptr());
-
-            // 4. Set entry's document state's origin to document's origin.
-            entry->document_state()->set_origin(document->origin());
+            entry->document_state()->set_document(document);
         }
 
-        // FIXME: 12. If entry's document state's request referrer is "client", then set it to request's referrer.
-        //     https://github.com/whatwg/html/issues/9767
-        //     What is "request" here?
-
-        // 13. If entry's document state's document is not null, then set entry's document state's ever populated to true.
+        // 7. If entry's document state's document is not null, then:
         if (entry->document()) {
+            // 1. Set entry's document state's ever populated to true.
             entry->document_state()->set_ever_populated(true);
+
+            // 2. If saveExtraDocumentState is true:
+            if (saveExtraDocumentState) {
+                // 1. Let document be entry's document state's document.
+                auto document = entry->document();
+
+                // 2. Set entry's document state's origin to document's origin.
+                entry->document_state()->set_origin(document->origin());
+
+                // FIXME: 3. If document's URL requires storing the policy container in history, then:
+            }
+            // FIXME: 3. If entry's document state's request referrer is "client", and navigationParams is a navigation params (i.e., neither null nor a non-fetch scheme navigation params), then:
+            //        https://github.com/whatwg/html/issues/9767
+            //        What is "request" here?
```

Comment:
curious that the confusing reference to "Request" was not addressed. given the issue was closed 8 months ago we should probably open a new bug.

Unless the new steps do lay out who "request" is in this context?

Reply:
Looking at it now, would *"entry's document state's request referrer"* refer to `entry->document_state()->request_referrer()`?  I've updated my last commit to try to implement this step and remove the FIXME

Reply:
I added this conditional statement, but left the following steps as `FIXME`'s, since it looks like we can't yet assert that navigationParams's request is not null, since navigationParams may be set to a `AK::StringView` when there are network errors.

I'll look into filling in those assertions in a later PR.